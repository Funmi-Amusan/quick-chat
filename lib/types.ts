export type ChatType = {
    id: number;
    name: string;
    lastMessage: string;
    lastMessageTime: string;
    avatar: string;
    isOnline: boolean;
    isTyping: boolean;
    messages: {
        id: number;
        sender: string;
        message: string;
        timestamp: string;
        isRead: boolean;
    };
};

export interface ChatData {
    createdAt: string;
    id: string;
    participants: string[];
    partner: string;
}

export interface UserData {
    id: string;
    email: string;
    username: string;
    chats: ChatData[];
}

export interface FormattedUser {
    id: string;
    email: string;
    username: string;
}

export interface FormattedChat {
    id: string;
    email: string;
    username: string;
}

export interface FirebaseMessage {
    id: string;
    content: string;
    senderId: string;
    timestamp: number;
    read: boolean;
}

export interface ChatPartner {
    id: string;
    username: string;
    isActive: boolean;
    isTyping: boolean;
    lastActive: number | null;
}