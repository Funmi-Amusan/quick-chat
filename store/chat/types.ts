import { Product } from "../product/types";
import { PaymentPlanType } from "@/enums/payment-plan-type.enum";
import { PaymentPlan } from "../util/types";
import { VoucherType } from "@/enums/voucher-type.enum";
import { DiscountType } from "@/enums/discount-type";
import { CartType } from "@/enums/cart-type.enum";
import { AppointmentType } from "@/enums/appointment-type.enum";
import { Operation } from "@/enums/operation-type.enum";
import { CartProduct } from "../dashboard/types";

export type InitialState = {
  loading: boolean;
  cartObject: CartObject;
  promoCodeData: PromoCodeData | null;
  paymentDetails: PaymentDetails | null;
};

export type SaveCartPayload = {
  product: string;
};

export type UsedPromoCodeDetails = {};

export type PromoCodeData = {
  hasPromoCode?: boolean;
  promoCodeError?: string;
  voucher?: {
    voucherType: VoucherType;
    discountType: DiscountType;
    discountValue: number;
    products: Product[];
    code: string;
  };
};

export type PaymentDetails = {};

export type PaymentPlanPrice = {
  paymentPlan: PaymentPlan;
  id: string;
  uniqueId: string;
  price: number;
  subscribeAndSave: number;
};

export type SubCartObject = {
  operation: Operation;
  items: CartProduct[];
  total: number;
};

export type CartObject = {
  cartType: CartType;
  cartAppointmentType: AppointmentType;
  paymentPlanType: PaymentPlanType;
  labTest: SubCartObject;
  kit: SubCartObject;
  vaccination: SubCartObject;
  medication: SubCartObject;
  consultation: SubCartObject;
};

export type PaymentPlanDropdownOption = {
  id: PaymentPlanType;
  name: string;
  label: string;
};
