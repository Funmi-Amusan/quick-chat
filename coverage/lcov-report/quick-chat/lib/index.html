
<!doctype html>
<html lang="en">

<head>
    <title>Code coverage report for quick-chat/lib</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <link rel="shortcut icon" type="image/x-icon" href="../../favicon.png" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
    
<body>
<div class='wrapper'>
    <div class='pad1'>
        <h1><a href="../../index.html">All files</a> quick-chat/lib</h1>
        <div class='clearfix'>
            
            <div class='fl pad1y space-right2'>
                <span class="strong">33.33% </span>
                <span class="quiet">Statements</span>
                <span class='fraction'>137/411</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">19.56% </span>
                <span class="quiet">Branches</span>
                <span class='fraction'>27/138</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">11.47% </span>
                <span class="quiet">Functions</span>
                <span class='fraction'>7/61</span>
            </div>
        
            
            <div class='fl pad1y space-right2'>
                <span class="strong">29.17% </span>
                <span class="quiet">Lines</span>
                <span class='fraction'>110/377</span>
            </div>
        
            
        </div>
        <p class="quiet">
            Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
        </p>
        <template id="filterTemplate">
            <div class="quiet">
                Filter:
                <input type="search" id="fileSearch">
            </div>
        </template>
    </div>
    <div class='status-line low'></div>
    <div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="data.ts"><a href="data.ts.html">data.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	</tr>

<tr>
	<td class="file high" data-value="firebase-config.ts"><a href="firebase-config.ts.html">firebase-config.ts</a></td>
	<td data-value="100" class="pic high">
	<div class="chart"><div class="cover-fill cover-full" style="width: 100%"></div><div class="cover-empty" style="width: 0%"></div></div>
	</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	</tr>

<tr>
	<td class="file low" data-value="firebase-sevice.ts"><a href="firebase-sevice.ts.html">firebase-sevice.ts</a></td>
	<td data-value="20.23" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 20%"></div><div class="cover-empty" style="width: 80%"></div></div>
	</td>
	<td data-value="20.23" class="pct low">20.23%</td>
	<td data-value="257" class="abs low">52/257</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="81" class="abs low">0/81</td>
	<td data-value="4.65" class="pct low">4.65%</td>
	<td data-value="43" class="abs low">2/43</td>
	<td data-value="15.25" class="pct low">15.25%</td>
	<td data-value="236" class="abs low">36/236</td>
	</tr>

<tr>
	<td class="file high" data-value="helpers.ts"><a href="helpers.ts.html">helpers.ts</a></td>
	<td data-value="97.4" class="pic high">
	<div class="chart"><div class="cover-fill" style="width: 97%"></div><div class="cover-empty" style="width: 3%"></div></div>
	</td>
	<td data-value="97.4" class="pct high">97.4%</td>
	<td data-value="77" class="abs high">75/77</td>
	<td data-value="84.37" class="pct high">84.37%</td>
	<td data-value="32" class="abs high">27/32</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="5" class="abs high">5/5</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="64" class="abs high">64/64</td>
	</tr>

<tr>
	<td class="file low" data-value="registerForPushNotificationAsync.ts"><a href="registerForPushNotificationAsync.ts.html">registerForPushNotificationAsync.ts</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="27" class="abs low">0/27</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="8" class="abs low">0/8</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="2" class="abs low">0/2</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="27" class="abs low">0/27</td>
	</tr>

<tr>
	<td class="file low" data-value="toast-config.tsx"><a href="toast-config.tsx.html">toast-config.tsx</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="11" class="abs low">0/11</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="6" class="abs low">0/6</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="11" class="abs low">0/11</td>
	</tr>

<tr>
	<td class="file low" data-value="useStorage.tsx"><a href="useStorage.tsx.html">useStorage.tsx</a></td>
	<td data-value="0" class="pic low">
	<div class="chart"><div class="cover-fill" style="width: 0%"></div><div class="cover-empty" style="width: 100%"></div></div>
	</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="29" class="abs low">0/29</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="11" class="abs low">0/11</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="7" class="abs low">0/7</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="29" class="abs low">0/29</td>
	</tr>

</tbody>
</table>
</div>
                <div class='push'></div><!-- for sticky footer -->
            </div><!-- /wrapper -->
            <div class='footer quiet pad2 space-top1 center small'>
                Code coverage generated by
                <a href="https://istanbul.js.org/" target="_blank" rel="noopener noreferrer">istanbul</a>
                at 2025-04-27T21:43:30.135Z
            </div>
        <script src="../../prettify.js"></script>
        <script>
            window.onload = function () {
                prettyPrint();
            };
        </script>
        <script src="../../sorter.js"></script>
        <script src="../../block-navigation.js"></script>
    </body>
</html>
    