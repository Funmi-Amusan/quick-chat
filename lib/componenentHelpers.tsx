import { Feather } from '@expo/vector-icons';
import { View, Text } from 'react-native';

import { LastMessageData } from './types';

export const generatePreviewMessageContent = (
  lastMessage: LastMessageData | undefined
): React.ReactNode | string => {
  if (!lastMessage) {
    return 'Be the first to send a message';
  }

  if (lastMessage?.messageType === 'text') {
    return lastMessage.content;
  } else if (lastMessage?.messageType === 'image') {
    return (
      <View className="flex-row items-center gap-2">
        <Feather name="camera" size={16} color="grey" />
        {lastMessage?.content && lastMessage?.content.length > 1 ? (
          <Text className="font-medium text-grey">{lastMessage?.content}</Text>
        ) : (
          <Text className="font-medium text-grey">Image</Text>
        )}
      </View>
    );
  } else if (lastMessage?.messageType === 'file') {
    return (
      <View className="flex-row items-center gap-2">
        <Feather name="file" size={16} color="grey" />
        {lastMessage?.content && lastMessage?.content.length > 1 ? (
          <Text className="font-medium text-grey">{lastMessage?.content}</Text>
        ) : (
          <Text className="font-medium text-grey">File</Text>
        )}
      </View>
    );
  } else {
    return '';
  }
};
