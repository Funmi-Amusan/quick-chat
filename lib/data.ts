export const shortDays = ['Sun', 'Mon', 'Tue', 'Wed', 'Thur', 'Fri', 'Sat'];
export const days = ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
export const emotionsData = [
    { range: [0, 10], hue: 270, label: 'Very Unpleasant' },
    { range: [10, 20], hue: 240, label: 'Unpleasant' },
    { range: [20, 30], hue: 120, label: 'Neutral' },
    { range: [30, 40], hue: 30, label: 'Pleasant' },
    { range: [40, 50], hue: 120, label: 'Very Pleasant' },
];
