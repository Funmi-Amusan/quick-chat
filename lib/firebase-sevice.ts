import {
  signInWithEmailAndPassword,
  createUserWithEmailAndPassword,
  signOut,
  updateProfile,
  User,
  UserCredential,
} from 'firebase/auth';
import {
  getDatabase,
  ref,
  onValue,
  push,
  update,
  get,
  serverTimestamp,
  Unsubscribe,
  onDisconnect,
  query,
  orderByChild,
  set,
  limitToLast,
  startAfter,
  endBefore,
} from 'firebase/database';
import {
  getStorage,
  ref as storageRef,
  uploadBytesResumable,
  getDownloadURL,
} from 'firebase/storage';
import React from 'react';
import { Alert } from 'react-native';
import Toast from 'react-native-toast-message';

import { auth } from './firebase-config';

import {
  ChatData,
  ChatPartner,
  FirebaseMessage,
  FormattedUser,
  ReplyMessageInfo,
  UserData,
} from '~/lib/types';

export interface FirebaseUserResponse {
  user: User;
}

const db = getDatabase();

export async function login(
  email: string,
  password: string
): Promise<FirebaseUserResponse | undefined> {
  try {
    const userCredential: UserCredential = await signInWithEmailAndPassword(auth, email, password);
    return { user: userCredential.user };
  } catch (e) {
    Toast.show({
      type: 'error',
      text1: 'Login failed',
      text2: 'Invalid email or password',
    });
    throw e;
  }
}

export async function logout(): Promise<void> {
  try {
    await signOut(auth);
  } catch (e) {
    console.error('[error logging out] ==>', e);
    throw e;
  }
}

export async function register(
  email: string,
  password: string,
  name?: string
): Promise<FirebaseUserResponse | undefined> {
  try {
    const userCredential = await createUserWithEmailAndPassword(auth, email, password);
    if (name) {
      await updateProfile(userCredential.user, { displayName: name });
    }
    return { user: userCredential.user };
  } catch (e) {
    console.error('[error registering] ==>', e);
    throw e;
  }
}

const formatFirebaseUser = (userId: string, userData: UserData): FormattedUser => {
  return {
    id: userId,
    email: userData.email,
    username: userData.username,
  };
};

export const listenToUserChats = (
  userId: string,
  callback: (chats: ChatData[], error?: Error) => void
): Unsubscribe => {
  const userChatsRef = ref(db, `users/${userId}/chatsSummary`);
  const unsubscribe = onValue(
    userChatsRef,
    async (snapshot) => {
      const chatIdsData = snapshot.val();
      if (chatIdsData) {
        const chatIds = Object.keys(chatIdsData);
        const chatPromises = chatIds.map(async (chatId) => {
          try {
            const chatSummaryRef = ref(db, `chatsSummary/${chatId}`);
            const chatSummarySnapshot = await get(chatSummaryRef);
            if (chatSummarySnapshot.exists()) {
              const chatData = chatSummarySnapshot.val();
              const participants = chatData.participantNames || {};
              const participantKeys = Object.keys(participants);
              const partnerKey = participantKeys.find((key) => key !== userId);
              let partnerName = 'Anonymous';
              const partnerId = '';
              if (partnerKey) {
                partnerName = participants[partnerKey];
              }
              return {
                id: chatId,
                lastMessage: chatData.lastMessage,
                partner: partnerName,
                partnerId,
                updatedAt: chatData.updatedAt,
              };
            }
            console.warn(`Chat data for ID ${chatId} not found.`);
            return null;
          } catch (err: any) {
            console.error(`Error fetching chat ${chatId}:`, err);
            return null;
          }
        });

        try {
          const chats = await Promise.all(chatPromises);
          const validChats = chats.filter((chat): chat is ChatData => chat !== null);
          callback(validChats);
        } catch (err: any) {
          console.error('Error processing chat promises:', err);
          callback([], err);
        }
      } else {
        callback([]);
      }
    },
    (error: Error) => {
      console.error('Error listening to user chats:', error);
      callback([], error);
    }
  );

  return unsubscribe;
};

export const fetchAllUsers = async (currentUserId: string): Promise<FormattedUser[]> => {
  const usersRef = ref(db, 'users');
  try {
    const snapshot = await get(usersRef);
    if (snapshot.exists()) {
      const usersData = snapshot.val() as Record<string, UserData>;
      const formattedUsers = Object.entries(usersData)
        .map(([userId, userData]) => formatFirebaseUser(userId, userData))
        .filter((user) => user.id !== currentUserId);
      return formattedUsers;
    } else {
      return [];
    }
  } catch (err: any) {
    console.error('Error fetching all users:', err);
    throw new Error(err.message || 'Failed to fetch users.');
  }
};

export const createChat = async (
  currentUserId: string,
  currentUserName: string,
  otherUserId: string,
  otherUserName: string
): Promise<string> => {
  const userIds = [currentUserId, otherUserId].sort();
  const chatId = userIds.join('_');
  const chatRef = ref(db, `chats/${chatId}`);
  const chatSummaryRef = ref(db, `chatsSummary/${chatId}`);

  const existingChat = await get(chatRef);
  if (existingChat.exists()) {
    Alert.alert('Chat already exists, returning existing chat ID:');
    return chatId;
  }

  const existingChatSummary = await get(chatSummaryRef);
  if (existingChatSummary.exists()) {
    console.error('Chat summary already exists, returning existing chat ID:');
    return chatId;
  }

  const participantsData = {
    [`${currentUserId}_${currentUserName}`]: {
      updatedAt: serverTimestamp(),
    },
    [`${otherUserId}_${otherUserName}`]: {
      updatedAt: serverTimestamp(),
    },
  };

  const participantNames = {
    [currentUserId]: currentUserName,
    [otherUserId]: otherUserName,
  };

  const newChatData = {
    participants: participantsData,
    createdAt: serverTimestamp(),
    updatedAt: serverTimestamp(),
    lastMessage: null,
    participantNames,
  };

  const updates: { [key: string]: any } = {};
  updates[`/chats/${chatId}`] = newChatData;
  updates[`/chatsSummary/${chatId}`] = {
    updatedAt: serverTimestamp(),
    lastMessage: null,
    participantNames,
  };
  updates[`/users/${currentUserId}/chatsSummary/${chatId}`] = true;
  updates[`/users/${otherUserId}/chatsSummary/${chatId}`] = true;

  try {
    await update(ref(db), updates);
    return chatId;
  } catch (error: any) {
    console.error('Error creating chat in service:', error);
    throw new Error(error.message || 'Failed to create chat.');
  }
};

export const listenToUserStatus = (
  userId: string,
  callback: (status: { online: boolean; last_seen: number } | null) => void
): Unsubscribe => {
  const userStatusRef = ref(db, `/status/${userId}`);
  return onValue(
    userStatusRef,
    (snapshot) => {
      if (snapshot.exists()) {
        callback(snapshot.val() as { online: boolean; last_seen: number });
      } else {
        callback(null);
      }
    },
    (error) => {
      console.error(`Error listening to status for user ${userId}:`, error);
      callback(null);
    }
  );
};

export const updateUserPresence = (userId: string, chatId: string, isActive: boolean) => {
  const userStatusRef = ref(db, `userStatus/${userId}`);
  const userChatStatusRef = ref(db, `chats/${chatId}/participants/${userId}`);
  const status = {
    isActive,
    lastActive: serverTimestamp(),
  };
  update(userStatusRef, status);
  update(userChatStatusRef, status);
};

export const setupPresenceDisconnectHandlers = (userId: string, chatId: string) => {
  const userStatusRef = ref(db, `userStatus/${userId}`);
  const userChatStatusRef = ref(db, `chats/${chatId}/participants/${userId}`);
  onDisconnect(userStatusRef).update({
    isActive: false,
    lastActive: serverTimestamp(),
  });
  onDisconnect(userChatStatusRef).update({
    isActive: false,
    lastActive: serverTimestamp(),
  });
};

export const setTypingStatus = (userId: string, chatId: string, isTyping: boolean) => {
  const typingRef = ref(db, `chats/${chatId}/participants/${userId}/isTyping`);
  update(typingRef, { isTyping });
};

export const markMessagesAsRead = (
  chatId: string,
  currentUserId: string,
  messages: FirebaseMessage[]
) => {
  const unreadMessages = messages.filter((msg) => msg.senderId !== currentUserId && !msg.read);
  if (unreadMessages.length) {
    const updates: Record<string, boolean> = {};
    unreadMessages.forEach((msg) => {
      updates[`chats/${chatId}/messages/${msg.id}/read`] = true;
    });
    update(ref(db), updates);
  }
};

export const fetchChatPartnerInfo = async (
  chatId: string,
  currentUserId: string
): Promise<ChatPartner> => {
  const chatMetaRef = ref(db, `chats/${chatId}`);

  try {
    const snapshot = await get(chatMetaRef);
    if (!snapshot.exists()) {
      throw new Error('Chat metadata not found.');
    }

    const chatData = snapshot.val();
    const participantIds = Object.keys(chatData.participants || {});
    const partnerIdWithName = participantIds.find((pId) => !pId.includes(currentUserId));

    let partnerId: string | undefined;
    if (partnerIdWithName?.includes('_')) {
      partnerId = partnerIdWithName?.split('_')[0];
    } else {
      partnerId = partnerIdWithName;
    }

    if (!partnerId) {
      throw new Error('Chat partner not found in participants.');
    }

    const userRef = ref(db, `users/${partnerId}`);
    const userSnapshot = await get(userRef);
    if (!userSnapshot || !userSnapshot.exists()) {
      throw new Error('Chat partner user data not found.');
    }
    const userData = userSnapshot.val();

    const statusRef = ref(db, `chats/${chatId}/participants/${partnerId}`);
    const statusSnapshot = await get(statusRef);
    const statusData = statusSnapshot.val() || {};

    const chatPartnerInfo: ChatPartner = {
      id: partnerId,
      username: userData.username,
      isActive: statusData.isActive || false,
      isTyping: statusData.isTyping || false,
      lastActive: statusData.lastActive || null,
      isLoggedIn: userData.isLoggedIn || false,
    };

    return chatPartnerInfo;
  } catch (error: any) {
    console.error('Error fetching chat partner info:', error);
    throw error;
  }
};

export const fetchInitialMessages = async (
  chatId: string,
  batchSize: number = 50
): Promise<FirebaseMessage[]> => {
  const messagesRef = ref(db, `chats/${chatId}/messages`);
  const initialMessagesQuery = query(
    messagesRef,
    orderByChild('timestamp'),
    limitToLast(batchSize)
  );
  try {
    const snapshot = await get(initialMessagesQuery);
    const messagesData = snapshot.val();
    if (messagesData) {
      const messagesList: FirebaseMessage[] = Object.entries(messagesData).map(
        ([key, value]: [string, any]) => ({
          id: key,
          content: value.content,
          imageUrl: value.imageUrl || null,
          senderId: value.senderId,
          timestamp: value.timestamp,
          read: value.read || false,
          reaction: value.reactions || undefined,
          replyMessage: value.replyMessage || null,
        })
      );
      return messagesList;
    } else {
      return [];
    }
  } catch (err) {
    console.error('Error fetching initial messages:', err);
    throw err;
  }
};

export const fetchOlderMessages = async (
  chatId: string,
  oldestTimestamp: number,
  batchSize: number = 50
): Promise<FirebaseMessage[]> => {
  const messagesRef = ref(db, `chats/${chatId}/messages`);
  const messagesQuery = query(
    messagesRef,
    orderByChild('timestamp'),
    endBefore(oldestTimestamp),
    limitToLast(batchSize)
  );

  try {
    const snapshot = await get(messagesQuery);
    const messagesData = snapshot.val();
    if (messagesData) {
      const messagesList: FirebaseMessage[] = Object.entries(messagesData).map(
        ([key, value]: [string, any]) => ({
          id: key,
          content: value.content,
          imageUrl: value.imageUrl || null,
          senderId: value.senderId,
          timestamp: value.timestamp,
          read: value.read || false,
          reaction: value.reactions || undefined,
          replyMessage: value.replyMessage || null,
        })
      );
      return messagesList;
    } else {
      return [];
    }
  } catch (err) {
    console.error('Error fetching older messages:', err);
    throw err;
  }
};

export const listenForNewMessages = (
  chatId: string,
  latestTimestamp: number,
  onNewMessage: (message: FirebaseMessage) => void,
  onError: (error: string | null) => void
) => {
  const messagesRef = ref(db, `chats/${chatId}/messages`);
  const newMessagesQuery = query(
    messagesRef,
    orderByChild('timestamp'),
    startAfter(latestTimestamp)
  );
  const unsubscribe = onValue(
    newMessagesQuery,
    (snapshot) => {
      const messagesData = snapshot.val();
      if (messagesData) {
        const newMessagesList: FirebaseMessage[] = Object.entries(messagesData).map(
          ([key, value]: [string, any]) => ({
            id: key,
            content: value.content,
            imageUrl: value.imageUrl || null,
            senderId: value.senderId,
            timestamp: value.timestamp,
            read: value.read || false,
            reaction: value.reactions || undefined,
            replyMessage: value.replyMessage || null,
          })
        );
        newMessagesList.sort((a, b) => b.timestamp - a.timestamp);
        newMessagesList.forEach((message) => {
          onNewMessage(message);
        });
      }
    },
    (err: any) => {
      console.error('Error listening for new messages:', err);
      onError(err.message || 'Failed to listen for new messages.');
    }
  );

  return unsubscribe;
};

export const sendMessage = async (
  chatId: string,
  userId: string,
  content: string,
  replyMessage: ReplyMessageInfo | null = null
) => {
  const messagesRef = ref(db, `chats/${chatId}/messages`);
  const chatMetaRef = ref(db, `chats/${chatId}`);
  const chatSummaryRef = ref(db, `chatsSummary/${chatId}`);
  const newMessageData = {
    content: content.trim(),
    senderId: userId,
    timestamp: serverTimestamp(),
    read: false,
    replyMessage: replyMessage ?? null,
  };
  try {
    await push(messagesRef, newMessageData);
    const lastMessageUpdate = {
      lastMessage: {
        content: content.trim(),
        timestamp: serverTimestamp(),
        senderId: userId,
        read: false,
      },
      updatedAt: serverTimestamp(),
    };
    await update(chatSummaryRef, lastMessageUpdate);
    await update(chatMetaRef, lastMessageUpdate);
  } catch (err: any) {
    console.error('Error sending message:', err);
    throw new Error('Failed to send message.');
  }
};

export const sendImageMessage = async (
  chatId: string,
  senderId: string,
  imageUri: string,
  text: string = '',
  replyTo: ReplyMessageInfo | null = null
) => {
  try {
    const storage = getStorage();
    const filename = imageUri.substring(imageUri.lastIndexOf('/') + 1);
    const storageReference = storageRef(storage, `chat_images/${chatId}/${Date.now()}_${filename}`);

    const response = await fetch(imageUri);
    const blob = await response.blob();
    const uploadTask = uploadBytesResumable(storageReference, blob);
    await new Promise((resolve, reject) => {
      uploadTask.on(
        'state_changed',
        (snapshot) => {
          const progress = (snapshot.bytesTransferred / snapshot.totalBytes) * 100;
          console.log('Upload is ' + progress + '% done');
        },
        (error) => {
          reject(error);
        },
        () => {
          resolve(null);
        }
      );
    });
    const downloadURL = await getDownloadURL(uploadTask.snapshot.ref);
    const messagesRef = ref(db, `chats/${chatId}/messages`);
    await push(messagesRef, {
      content: text,
      imageUrl: downloadURL,
      senderId,
      timestamp: serverTimestamp(),
      read: {},
      replyTo: replyTo
        ? {
            id: replyTo.id,
            content: replyTo.content,
            senderId: replyTo.senderId,
          }
        : null,
    });
    const chatMetaRef = ref(db, `chats/${chatId}`);
    await push(chatMetaRef, {
      lastMessage: {
        content: text ? text : imageUri,
        senderId,
        timestamp: serverTimestamp(),
        hasImage: true,
      },
      updatedAt: serverTimestamp(),
    });
    return true;
  } catch (error) {
    console.error('Error uploading image:', error);
    throw error;
  }
};

export const resetTypingStatus = (userId: string, chatId: string) => {
  const typingRef = ref(db, `chats/${chatId}/participants/${userId}/isTyping`);
  update(typingRef, { isTyping: false });
};

export const reactToMessage = async (
  messageId: string,
  emoji: string,
  chatId: string
  // currentUserId: string .... if i want to do a group chat, this will be important
) => {
  const reactionRef = ref(db, `chats/${chatId}/messages/${messageId}/reactions`);
  await set(reactionRef, emoji);
};
