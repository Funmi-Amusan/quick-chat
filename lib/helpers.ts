import { FormattedUser, ChatData, UserData, FormattedChat } from './types';

export const formatUserData = (userData: UserData[]): FormattedUser[] => {
    const formattedData = [];

    for (const userId in userData) {
        if (userData.hasOwnProperty(userId)) {
            const user = userData[userId];
            formattedData.push({
                id: userId,
                email: user.email,
                username: user.username,
            });
        }
    }

    return formattedData;
};

export const formatTimestamp = (timestamp: string) => {
    const date = new Date(+timestamp);
    let hours = date.getHours();
    const minutes = date.getMinutes();
    const ampm = hours >= 12 ? 'pm' : 'am';

    hours = hours % 12 || 12;
    const formattedMinutes = minutes < 10 ? `0${minutes}` : minutes;

    return `${hours}:${formattedMinutes}${ampm}`;
};

export const formatMomentAgo = (timestamp: number | null) => {
    if (!timestamp) return '';

    const date = new Date(timestamp);
    const now = new Date();
    const diffMinutes = Math.floor((now.getTime() - date.getTime()) / (1000 * 60));

    if (diffMinutes < 1) return 'just now';
    if (diffMinutes < 60) return `${diffMinutes}m ago`;

    const diffHours = Math.floor(diffMinutes / 60);
    if (diffHours < 24) return `${diffHours}h ago`;

    return date.toLocaleDateString();
};
