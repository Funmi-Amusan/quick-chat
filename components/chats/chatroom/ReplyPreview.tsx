import Ionicons from '@expo/vector-icons/Ionicons';
import MaterialCommunityIcons from '@expo/vector-icons/MaterialCommunityIcons';
import React from 'react';
import { View, Text, Image } from 'react-native';

import { ReplyMessageInfo } from '~/lib/types';

const ReplyPreview = ({
  replyMessage,
  setReplyMessage,
}: {
  replyMessage: ReplyMessageInfo;
  setReplyMessage: (arg0: ReplyMessageInfo | null) => void;
}) => {
  return (
    <View className="h-12 flex-row items-center gap-2 border-l-4 border-mint bg-black/20 dark:bg-white/20 ">
      <View className="flex-grow flex-row items-start gap-2 px-2">
        {replyMessage.imageUrl && <MaterialCommunityIcons name="camera" size={20} color="grey" />}
        <Text className=" line-clamp-1 text-sm text-greyText-light dark:text-greyText-dark">
          {replyMessage.content ? replyMessage.content : replyMessage.imageUrl ? 'Photo' : ''}
        </Text>
      </View>
      <View className=" my-1 flex-row items-center justify-center">
        <Image
          source={{ uri: replyMessage.imageUrl || '' }}
          className="aspect-square h-full rounded-md"
          resizeMode="cover"
        />
        <Ionicons
          name="close-circle-outline"
          size={24}
          color="white"
          className="mx-2"
          onPress={() => setReplyMessage(null)}
        />
      </View>
    </View>
  );
};

export default ReplyPreview;
