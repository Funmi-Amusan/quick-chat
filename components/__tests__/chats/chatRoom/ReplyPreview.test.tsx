import { render, screen, fireEvent } from '@testing-library/react-native';
import React from 'react';
import ReplyPreview from '~/components/chats/chatRoom/ReplyPreview';
import { ReplyMessageInfo } from '~/lib/types';

test('displays ReplyPreview component correctly', async () => {
  const replyMessage: ReplyMessageInfo = {
    id: 'string',
    content: 'string',
    senderId: 'string',
    imageUrl: null,
  };

  const replyPreviewComponent = render(
    <ReplyPreview replyMessage={replyMessage} setReplyMessage={() => {}} />
  );

  expect(replyPreviewComponent.toJSON()).toMatchSnapshot();
});
