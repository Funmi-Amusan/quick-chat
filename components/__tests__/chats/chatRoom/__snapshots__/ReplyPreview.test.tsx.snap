// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays ReplyPreview component correctly 1`] = `
<View
  className="h-12 flex-row items-center gap-2 border-l-4 border-mint bg-black/20 dark:bg-white/20 "
>
  <View
    className="flex-grow flex-row items-start gap-2 px-2"
  >
    <Text
      className=" line-clamp-1 text-sm text-greyText-light dark:text-greyText-dark"
    >
      string
    </Text>
  </View>
  <View
    className=" my-1 flex-row items-center justify-center"
  >
    <Image
      className="aspect-square h-full rounded-md"
      resizeMode="cover"
      source={
        {
          "uri": "",
        }
      }
    />
    <Text
      allowFontScaling={false}
      className="mx-2"
      onPress={[Function]}
      selectable={false}
      style={
        [
          {
            "color": "white",
            "fontSize": 24,
          },
          undefined,
          {
            "fontFamily": "ionicons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {},
        ]
      }
    >
      Ó≠ç
    </Text>
  </View>
</View>
`;
